active/0002-centalized-mgmt-resource-types-provider-versions.md

```diff
@@ -0,0 +1,247 @@
+
+1. The proposal uses `bicepconfig.json` to declare the provider information and is ruled by `bicepconfig.json` merge semantics. The merge semantics for `bicepconfig.json` specify that for a given Bicep source file, the configuration file closest to the Bicep file in the directory hierarchy is used. This behavior is not intuitive and may cause users that have specified multiple config files in their project file-structure to experience unexpected behaviors that may include build errors when Bicep attempts to resolve a namespace identifier. This is illustrated by the example below.
+
+**Example: Provider resolution error due as a result of `bicepconfig.json` merge semantics**
+
+Consider a Bicep project under folder `src` with the following files:
+
+- `src/main.bicep`
+
+```bicep
+provider foo as mainFooProvider
+```
+
+- `src/bicepconfig.json`
+
+```bicep
+{
+    "providers":{
+        "foo": {
+            "source": {
+                "registry": "mcr.microsoft.com/bicep/providers/az",
+                "version": "1.2.3",
+            },
+            "isDefaultImport": false
+        },
+        "bar" : {
+            "source": {
+                "registry": "mcr.microsoft.com",
+                "repository": "bicep/providers",
+                "name":
+                "version": "3.2.1"
+            },
+            "isDefaultImport": false
+        }
+    }
+}
+```
+
+- `src/modules/mod.bicep`
+
+```bicep
+provider bar as modBarProvider
+provider foo as modFooProvider
+```
+
+- `src/modules/bicepconfig.json`
+
+```bicep
+{
+     "experimentalFeaturesEnabled": {
+        "compileTimeImports": false
+     },
+}
+```
+
+The user may intuitively expect Bicep to resolve the provider details of `modFooProvider` by inspection of `src/bicepconfig.json` and for this project to compile successfully in a way that feels familiar with other programming languages where merge semantics are applied on code dependencies (e.g. NPMs `package.json`), however, Bicep config merge semantics uses the configuration file closest to the Bicep file in the directory hierarchy and merges them to the builtin config.
+
+In this case Bicep will try to resolve the `modFooProvider` details from `src/modules/bicepconfig.json`. The config doesn't specify a `providers` section, and the builtin config doesn't include a definition for provider `foo` and so the compiler cannot resolve the provider declaration statement and will emit a diagnostic.
+
+If we replace `foo` with `az` in the example above, Bicep resolves `modFooProvider`'s version to be the version of `az` specified in the builtin configuration and not the version `1.2.3` specified in `src/modules/mod.bicep` (inherited from `src/bicepconfig.json`) which is also not intuitive. The reson for this behavior is that current `bicepconfig.json` merge semantics merge `src/modules/bicepconfig.json` with the builtin config and the version applied would be the value specified in the builtin az provider version.
```

Comment:
> also not intuitive

This is an opinion. You could argue that "the closest bicepconfig.json is used" is intuitive and simpler to reason about.

Reply:
replaced intuitive with expected